### PR TITLE
Stravid87/more tests v2

### PR DESCRIPTION
**Pull request by stravid87**
Changes: 
1) Pre checks added to the function internals/main.go/NewClient
2) Method, ```GetURL() string``` added to the interface `ClientImpl`
3) Simple test cases added to main_test.go

### CODE REVIEW

**Comment by stravid87** [comment]
Ahh.... I need to test locally....

**Comment by stravid87** [comment]
Local testing now complete.

File: internals/main.go [hunk 1]
```diff
@@ -49,10 +74,23 @@ func (c *Client) Do(url string, req *utils.Request, sharedSecret *utils.JWK, isS
...
-// TODO: Daniel, can client.transfer and client.do be combined?
-// If not, why keep them separate?
-// transfer sends the request to the remote server through the layer8 proxy server
+// Performs Prechecks and then transforms the byte slice to a utils.Response struct.
 func (c *Client) transfer(sharedSecret *utils.JWK, req *utils.Request, url string, isStatic bool, UpJWT, UUID string) (*utils.Response, error) {
+	// Prechecks
+	if sharedSecret == nil || req == nil {
+		return &utils.Response{
+			Status:     400,
+			StatusText: "client.transfer(...) error. Malformed url, UpJWT, or UUID",
```
Comment (DeeStarks) [inline internals/main.go:83]: We can make the status text more descriptive of the error like including that the shared secret or the request is not valid, or we can log it for easy debugging

File: internals/main.go [hunk 2]
```diff
@@ -49,10 +74,23 @@ func (c *Client) Do(url string, req *utils.Request, sharedSecret *utils.JWK, isS
...
+		}, nil
+	}
+
+	if url == "" || UpJWT == "" || UUID == "" {
+		return &utils.Response{
+			Status:     400,
+			StatusText: "client.transfer(...) error. Malformed url, UpJWT, or UUID",
```
Comment (DeeStarks) [inline internals/main.go:90]: This as well can be more descriptive

File: internals/main_test.go [hunk 3]
```diff
@@ -186,17 +219,82 @@ func TestClientDo(t *testing.T) {
...
+				res := client.Do(RequestURL, malformedReq, symmkey, false, up_JWT, uuid)
+				assert.NotNil(t, res)
+				assert.Equal(t, tc.Expected, res.Status) // 200, 500
+				for k, v := range ResponseHeader {
+					assert.Equal(t, v, res.Headers[k])
+				}
+				assert.Equal(t, ResponsePayload, res.Body)
+				assert.Equal(t, http.StatusText(ResponseStatusCode), res.StatusText)
+			})
+		}
```
Comment (DeeStarks) [inline internals/main_test.go:298]: This is clean and elegant!

File: internals/main.go (hunk 1 above)
Comment (stravid87) [reply internals/main.go:83]: Good call. Let me do that now.

File: internals/main_test.go (hunk 3 above)
Comment (stravid87) [reply internals/main_test.go:298]: Thanks! As the junior Go developer, it's very encouraging to earn the compliment.

**Review by stravid87 — COMMENTED** [review]
Responding to Daniel's comments and suggestions.

**Pull request merged**